cpu/nrf52/radio/nrf802154/nrf802154_radio.c

```diff
@@ -0,0 +1,703 @@
+#include <string.h>
+#include <errno.h>
+
+#include "cpu.h"
+#include "luid.h"
+
+#include "net/ieee802154.h"
+#include "periph/timer.h"
+#include "nrf802154.h"
+#include "net/ieee802154/radio.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define ED_RSSISCALE        (4U)
+#define ED_RSSIOFFS         (-92)
+
+#define LIFS                (40U)
+#define SIFS                (12U)
+#define SIFS_MAXPKTSIZE     (18U)
+#define TIMER_FREQ          (62500UL)
+
+#define TX_POWER_MIN        (-40)
+#define TX_POWER_MAX        (8)
+
+#define DEFAULT_SHORTS      (RADIO_SHORTS_RXREADY_START_Msk | \
+                             RADIO_SHORTS_TXREADY_START_Msk);
+
+#define CCA_SHORTS          (RADIO_SHORTS_RXREADY_CCASTART_Msk | \
+                             RADIO_SHORTS_CCAIDLE_STOP_Msk | \
+                             RADIO_SHORTS_CCAIDLE_TXEN_Msk | \
+                             RADIO_SHORTS_CCABUSY_DISABLE_Msk | \
+                             RADIO_SHORTS_TXREADY_START_Msk);
+
+#define ACK_LENGTH          (5U)
+
+#define MAC_TIMER_CHAN_ACK  (0U)
+#define MAC_TIMER_CHAN_IFS  (1U)
+
+#if IS_USED(MODULE_IEEE802154_RADIO_HAL)
+static uint8_t rxbuf[IEEE802154_FRAME_LEN_MAX + 3]; /* len PHR + PSDU + LQI */
+static uint8_t txbuf[IEEE802154_FRAME_LEN_MAX + 3]; /* len PHR + PSDU + LQI */
+static uint8_t ack[ACK_LENGTH];
+
+typedef enum {
+    STATE_IDLE,
+    STATE_TX,
+    STATE_ACK,
+    STATE_RX,
+    STATE_CCA_CLEAR,
+    STATE_CCA_BUSY,
+} nrf802154_state_t;
+
+static volatile uint8_t _state;
+
+static uint8_t nrf802154_short_addr[IEEE802154_SHORT_ADDRESS_LEN];
+static uint8_t nrf802154_long_addr[IEEE802154_LONG_ADDRESS_LEN];
+static uint16_t nrf802154_pan_id;
+
+static bool ifs;
+static bool cca_send = true;
+static bool ack_filter = true;
+static bool promisc;
+
+static const ieee802154_radio_ops_t nrf802154_ops;
+
+ieee802154_dev_t nrf802154_hal_dev = {
+    .driver = &nrf802154_ops,
+};
+
+static bool _l2filter(uint8_t *mhr)
+{
+    uint8_t dst_addr[IEEE802154_LONG_ADDRESS_LEN];
+    le_uint16_t dst_pan;
+    uint8_t pan_bcast[] = IEEE802154_PANID_BCAST;
+
+    int addr_len = ieee802154_get_dst(mhr, dst_addr, &dst_pan);
+
+    /* filter PAN ID */
+    /* Will only work on little endian platform (all?) */
+
+    if ((memcmp(pan_bcast, dst_pan.u8, 2) != 0) &&
+        (memcmp(&nrf802154_pan_id, dst_pan.u8, 2) != 0)) {
+        return false;
+    }
+
+    /* check destination address */
+    if (((addr_len == IEEE802154_SHORT_ADDRESS_LEN) &&
+          (memcmp(nrf802154_short_addr, dst_addr, addr_len) == 0 ||
+           memcmp(ieee802154_addr_bcast, dst_addr, addr_len) == 0)) ||
+        ((addr_len == IEEE802154_LONG_ADDRESS_LEN) &&
+          (memcmp(nrf802154_long_addr, dst_addr, addr_len) == 0))) {
```

Comment:
Nope, there's only with short addresses. From the standard:
```
— The Frame Type subfield shall not contain a reserved frame type.
— The Frame Version subfield shall not contain a reserved value.
— If a destination PAN identifier is included in the frame, it shall match macPANId or shall be the
broadcast PAN identifier (0xffff).
— If a short destination address is included in the frame, it shall match either macShortAddress or the
broadcast address (0xffff). Otherwise, if an extended destination address is included in the frame, it
shall match aExtendedAddress.
— If the frame type indicates that the frame is a beacon frame, the source PAN identifier shall match
macPANId unless macPANId is equal to 0xffff, in which case the beacon frame shall be accepted
regardless of the source PAN identifier.
— If only source addressing fields are included in a data or MAC command frame, the frame shall be
accepted only if the device is the PAN coordinator and the source PAN identifier matches
macPANId.
```

Reply:
Huh now I wonder how `ping ff02::1` works.